File: 프로그래머스/4/1843. 사칙연산/사칙연산.py
import math

def solution(arr):
    nums = [int(x) for x in arr[::2]]
    ops = arr[1::2]
    len_nums = len(nums)

    max_dp = [[-math.inf for _ in range(len_nums)] for _ in range(len_nums)]
    min_dp = [[math.inf for _ in range(len_nums)] for _ in range(len_nums)]

    for i in range(len_nums):
        max_dp[i][i] = nums[i]
        min_dp[i][i] = nums[i]

    for s in range(1, len_nums):
        for i in range(len_nums - s):
            j = s + i

            for k in range(i, j):
                if ops[k] == '+':
                    max_dp[i][j] = max(max_dp[i][j], max_dp[i][k] + max_dp[k + 1][j])
                    min_dp[i][j] = min(min_dp[i][j], min_dp[i][k] + min_dp[k + 1][j])
                else:
                    max_dp[i][j] = max(max_dp[i][j], max_dp[i][k] - min_dp[k + 1][j])
                    min_dp[i][j] = min(min_dp[i][j], min_dp[i][k] - max_dp[k + 1][j])

    return max_dp[0][len_nums - 1]

File: 프로그래머스/4/1843. 사칙연산/사칙연산.java
import java.util.*;

class Solution {
    public int solution(String arr[]) {
        ArrayList<Integer> nums = new ArrayList<>();
        ArrayList<String> ops = new ArrayList<>();

        for (int i = 0; i < arr.length; i++) {
            if (i % 2 == 0) {
                nums.add(Integer.parseInt(arr[i]));
            } else {
                ops.add(arr[i]);
            }
        }

        int len = nums.size();
        int[][] maxArr = new int[len][len];
        int[][] minArr = new int[len][len];

        for (int i = 0; i < len; i++) {
            Arrays.fill(maxArr[i], Integer.MIN_VALUE);
            Arrays.fill(minArr[i], Integer.MAX_VALUE);
            maxArr[i][i] = minArr[i][i] = nums.get(i);
        }

        for (int s = 1; s < len; s++) {
            for (int i = 0; i < len - s; i++) {
                int j = s + i;

                for (int k = i; k < j; k++) {
                    if (ops.get(k).equals("+")) {
                        maxArr[i][j] = Math.max(maxArr[i][j], maxArr[i][k] + maxArr[k + 1][j]);
                        minArr[i][j] = Math.min(minArr[i][j], minArr[i][k] + minArr[k + 1][j]);
                    } else {
                        maxArr[i][j] = Math.max(maxArr[i][j], maxArr[i][k] - minArr[k + 1][j]);
                        minArr[i][j] = Math.min(minArr[i][j], minArr[i][k] - maxArr[k + 1][j]);
                    }
                }
            }
        }

        return maxArr[0][len - 1];
    }
}